File: app/src/main/java/org/techtown/samplerecorder/Main/AudioRecord.java
package org.techtown.samplerecorder.Main;

import android.content.Context;
import android.media.AudioFormat;
import android.media.MediaRecorder;
import android.util.Log;
import android.widget.Toast;

import java.io.File;
import java.io.FileNotFoundException;
import java.io.FileOutputStream;
import java.io.IOException;
import java.nio.ByteBuffer;
import java.nio.ByteOrder;
import java.text.SimpleDateFormat;
import java.util.Date;

public class AudioRecord {

    public static int dataMax;

    private android.media.AudioRecord audioRecord = null;
    private Thread recordThread = null;
    private FileOutputStream outputStream;
    private File file;
    private byte[] audioData = null;
    private int record_bufferSize, len_audioData;

    public void init(int bufferSize) {
//        myLog.d("method activate");

        audioData = null;
        record_bufferSize = bufferSize;
    }

    public void start(int source, int channel, int sampleRate, Queue queue, boolean fileDrop) {
//        myLog.d("method activate");

        if(audioRecord == null) {
            audioRecord = new android.media.AudioRecord(
                    source,
                    sampleRate,
                    channel,
                    AudioFormat.ENCODING_PCM_16BIT,
                    record_bufferSize
            );
        }

        if (fileDrop) {
//            file = new File("/mnt/sdcard/audioDrop/", fileName(System.currentTimeMillis()));
            file = new File("/data/user/0/org.techtown.samplerecorder/file/", fileName(System.currentTimeMillis()));
            myLog.d(file.getAbsolutePath());
            outputStream = null;

            try {
                outputStream = new FileOutputStream(file);  // TODO 여기서 생성이 안 되고, outputStream이 계속 null 발생
            } catch (FileNotFoundException e) {
                e.printStackTrace();
            }
        }

        audioRecord.startRecording();

        recordThread = new Thread(new Runnable() {
            @Override
            public void run() {

                len_audioData = 0;
                while(MainActivity.isRecording) {
                    audioData = new byte[record_bufferSize];  // prevent from overwritting data
                    len_audioData = audioRecord.read(audioData, 0, record_bufferSize);  // audioRecord -> audioData
                    queue.enqueue(audioData);

                    // using draw waveform in MainActivity
                    dataMax = 0;
                    for (int i = 0; i < audioData.length; i++) {
                        ByteBuffer buffer = ByteBuffer.wrap(audioData);
                        buffer.order(ByteOrder.LITTLE_ENDIAN);
                        dataMax = 10 * Math.abs(buffer.getShort());
                    }

                    if (fileDrop) {
                        try {
                            if (outputStream != null) {
                                myLog.d("드롭 파일 생성 중22");
                                outputStream.write(audioData, 0, len_audioData);
                            }
                        } catch (IOException e) {
                            e.printStackTrace();
                        }
                    }

                }
            }
        });

        recordThread.start();
    }

    public void stop() {
//        myLog.d("method activate");

        if (audioRecord != null) {
            if (audioRecord.getState() != android.media.AudioRecord.RECORDSTATE_STOPPED) {
                try {
                    audioRecord.stop();
                } catch (IllegalStateException e) {
                    e.printStackTrace();
                }

                audioRecord.release();
                audioRecord = null;
                recordThread = null;
            }
        }
    }

    public void release(Context context, boolean fileDrop) {
//        myLog.d("method activate");

        if (fileDrop) {
            try {
                outputStream.flush();  // TODO : null reference issue
                outputStream.close();
            } catch (IOException e) {
                myLog.d("exception while closing output stream " + e.toString());
                e.printStackTrace();
            }

            Toast.makeText(context, file.getAbsolutePath() + " 저장 완료", Toast.LENGTH_LONG).show();
        }
    }

    public String fileName(long realtime) {
        Date date = new Date(realtime);
        SimpleDateFormat dateFormat = new SimpleDateFormat("yyyy.MM.dd_HH_mm_ss");
        return dateFormat.format(date) + ".pcm";
    }
}

File: app/src/main/java/org/techtown/samplerecorder/Main/AudioRecord.kt
package org.techtown.samplerecorder.Main

import android.annotation.SuppressLint
import android.content.Context
import android.media.AudioFormat
import android.media.AudioRecord
import android.widget.Toast
import org.techtown.samplerecorder.MainActivity.Companion.bufferSize
import org.techtown.samplerecorder.MainActivity.Companion.filePath
import org.techtown.samplerecorder.MainActivity.Companion.isRecording
import org.techtown.samplerecorder.MainActivity.Companion.recordChannel
import org.techtown.samplerecorder.MainActivity.Companion.recordRate
import org.techtown.samplerecorder.MainActivity.Companion.source
import java.io.File
import java.io.FileNotFoundException
import java.io.FileOutputStream
import java.io.IOException
import java.nio.ByteBuffer
import java.nio.ByteOrder
import java.text.SimpleDateFormat
import java.util.*
import kotlin.math.abs

@Suppress("NULLABILITY_MISMATCH_BASED_ON_JAVA_ANNOTATIONS")
class AudioRecord {
    private val TAG = this.javaClass.simpleName

    private var audioRecord: AudioRecord? = null
    private var recordThread: Thread? = null
    private var outputStream: FileOutputStream? = null
    private var file: File? = null

    fun start(queue: Queue, fileDrop: Boolean) {  // TODO 함수로 다 쪼개기
        if (audioRecord == null) {
            audioRecord = AudioRecord(
                source,
                recordRate,
                recordChannel,
                AudioFormat.ENCODING_PCM_16BIT,
                bufferSize
            )
        }
        var audioData: ByteArray?

        if (fileDrop) {
            file = File(filePath, fileName())
            outputStream = null
            try { outputStream = FileOutputStream(file) }
            catch (e: FileNotFoundException) { e.printStackTrace() }
        }
        
        audioRecord!!.startRecording()
        
        recordThread = Thread {  // TODO Coroutine
            var dataSize: Int

            while (isRecording) {
                audioData = ByteArray(bufferSize) // prevent from overwritting data
                dataSize = audioRecord!!.read(
                    audioData!!,
                    0,
                    bufferSize
                ) // audioRecord -> audioData
                queue.enqueue(audioData)

                // using draw waveform in MainActivity
                dataMax = 0
                for (i in audioData!!.indices) {
                    val buffer = ByteBuffer.wrap(audioData)
                    buffer.order(ByteOrder.LITTLE_ENDIAN)
                    dataMax = 10 * abs(buffer.short.toInt())
                }
                
                if (fileDrop) {
                    try {
                        if (outputStream != null) {
                            outputStream!!.write(audioData, 0, dataSize)
                        }
                    } catch (e: IOException) {
                        e.printStackTrace()
                    }
                }
            }
        }
        recordThread!!.start()
    }

    fun stop() {  // TODO Merge
        if (audioRecord != null) {
            if (audioRecord!!.state != AudioRecord.RECORDSTATE_STOPPED) {
                try {
                    audioRecord!!.stop()
                } catch (e: IllegalStateException) {
                    e.printStackTrace()
                }
                audioRecord!!.release()
                audioRecord = null
                recordThread = null
            }
        }
    }

    fun release(context: Context?, fileDrop: Boolean) {
        if (fileDrop) {
            try {
                outputStream!!.flush() // TODO : null reference issue
                outputStream!!.close()
            } catch (e: IOException) {
                myLog.d("exception while closing output stream $e")
                e.printStackTrace()
            }
            Toast.makeText(context, file!!.absolutePath + " 저장 완료", Toast.LENGTH_LONG).show()
        }
    }

    @SuppressLint("SimpleDateFormat")
    private fun fileName(): String {
        val date = Date(System.currentTimeMillis())
        val dateFormat = SimpleDateFormat("yyyy.MM.dd HH:mm:ss")
        return dateFormat.format(date) + ".pcm"
    }

    companion object {
        var dataMax = 0
    }
}